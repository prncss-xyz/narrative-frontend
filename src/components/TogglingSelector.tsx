function add<T>(state: T[], value: T) {
  return [...state, value];
}

function remove<T>(state: T[], value: T) {
  return state.filter((value_) => value !== value_);
}

interface Item<T> {
  key: T;
  toElem: ({
    active,
    disabled,
    onClick,
  }: {
    active: boolean;
    disabled?: boolean;
    onClick: React.MouseEventHandler;
  }) => React.ReactNode;
}

export function TogglingSelector<T>({
  disabled,
  state,
  setState,
  items,
}: {
  state: T[];
  setState: (state: T[]) => void;
  disabled?: boolean;
  items: Item<T>[];
}) {
  return (
    <>
      {items.map(({ key, toElem }) => {
        const active = state.includes(key);
        return (
          <div key={String(key)}>
            {toElem({
              active,
              disabled,
              onClick: () =>
                setState(active ? remove(state, key) : add(state, key)),
            })}
          </div>
        );
      })}
    </>
  );
}
