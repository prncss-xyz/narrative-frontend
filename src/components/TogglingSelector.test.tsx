import { fireEvent, render, screen } from "@testing-library/react";
import { TogglingSelector } from "./TogglingSelector";

describe("TogglingSelector", () => {
  it("toggle key appartenance to list when clicked on associated element", () => {
    const keys = ["a", "b", "c"];
    const state = ["a", "b"];
    const setState = vi.fn();
    const { container } = render(
      <TogglingSelector
        state={state}
        setState={setState}
        items={keys.map((key) => ({
          key,
          toElem: ({ onClick }) => <button onClick={onClick}>{key}</button>,
        }))}
      />
    );
    expect(container).toMatchSnapshot();
    fireEvent.click(screen.getByText("b"));
    expect(setState.mock.calls[0][0]).toEqual(["a"]);
    fireEvent.click(screen.getByText("c"));
    expect(setState.mock.calls[1][0]).toEqual(["a", "b", "c"]);
  });
});
