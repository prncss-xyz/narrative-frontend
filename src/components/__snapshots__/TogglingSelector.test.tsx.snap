// Vitest Snapshot v1

exports[`TogglingSelector > toggle key appartenance to list when clicked on associated element 1`] = `
<div>
  <div>
    <button>
      a
    </button>
  </div>
  <div>
    <button>
      b
    </button>
  </div>
  <div>
    <button>
      c
    </button>
  </div>
</div>
`;
